### ISSUE TITLE
OidcReturnUrlParser.IsValidReturnUrl fails if return URL is not local

### ISSUE THREAD

**Issue by lwansbrough** [IdentityServer/IdentityServer4#2594]
My UI is rendered via a Node.js server running on port 8080. This UI includes the consent and sign in flows. A .NET server runs on 50291 which is where IS4 and my API are located.

In order to accomplish this, I've set the following URIs in my service options:

`BaseUri` is `http://localhost:50291`, while `ClientUri` is `http://localhost:8080`
```
options.IssuerUri = Configuration["BaseUri"];
options.PublicOrigin = Configuration["BaseUri"];
options.UserInteraction.ErrorUrl = $"{Configuration["ClientUri"]}/auth/error";
options.UserInteraction.ConsentUrl = $"{Configuration["ClientUri"]}/auth/consent";
options.UserInteraction.LoginUrl = $"{Configuration["ClientUri"]}/auth/signin";
options.UserInteraction.LogoutUrl = $"{Configuration["ClientUri"]}/auth/signout";
```
Visiting `http://localhost:50291/connect/authorize` with the correct params redirects me to `http://localhost:8080/auth/signin?returnUrl...` as expected. The return URL starts with `http://localhost:50291` as expected, because it is on a separate domain. When the UserInteraction URIs are set to local paths, the return URL starts with `/connect` -- a local path.

When I sign in, my code runs the following line of code, which when running the UI from the same domain as IS4, works as expected.

`var context = await interaction.GetAuthorizationContextAsync(returnUrl);`

However, when running the UI on a separate domain, context is null (returnUrl in this case would start with `http://localhost:50291`). I believe this is because `OidcReturnUrlParser.IsValidReturnUrl` returns false (my return URL is not a local path, which is required in this method's implementation.)

My question is whether or not this is intended behaviour, and I've done something wrong, or if it's a bug.

**Comment by lwansbrough** [comment]
Thanks for the reply. That clarifies the issue for me. I can provide a little more insight into how my app works, it it helps.

Everything data related happens on the “api.” subdomain. That includes my own cookie auth, as well as IS4. The app is on www. Sign in is handled through an API endpoint. That means authentication state is local to the IS4 application. For consent and error information, I have created two additional API endpoints which return the current consent/error state of the authenticated user. So that’s how I get that information to the client side. Depending on where we are in the flow, IS4 may need to redirect the client. In this case, if the current request is made via XHR, we can’t simply do a 302. Instead we send back a response which indicates the client needs to redirect itself to the URL provided.

I’m mostly fine leaving my shim code in for now. I check if the URL has my domain for its host and replace the URL with just the path if it does.

But yeah, I think while the flow is complicated to implement, there is a need if someone is trying to implement the entire flow inside their SPA which is on a different server.

**Comment by brockallen** [comment]
And at some point if the demand is high enough we might eventually finish out the implementation to allow for the UI to be on a diff domain than the STS.

**Comment by mikagouzee** [comment]
Hi

I believe my issue is the same.

I'm trying to implement an implicit grant type since I have a vueJs spa as front-end. I know your documentation advises to make it via authorization code , but since there's no documentation for that yet, I wanted to try implicit first.

I've set things up as in this documentation : http://docs.identityserver.io/en/latest/quickstarts/6_javascript_client.html#refjavascriptquickstart ; the only difference worth mentionning being that the front is served by node.

When I call userManager.signinredirect, I'm send on {myIdentityServerBackend}/Account/login?ReturnUrl=%2Fconnect%2Fauthorize%2Fcallback%2F .... 

I don't know if the issue is the same or not, but it definitely looks similar enough for me to ask.

Let me know if I should provide more info, if I should do anything differently, or if you want me to go back to my cave and never get out ever again.

**Comment by vincent-polite** [comment]
Hello, it is Aug 2019, and I was wondering if the UserInteraction options were still intended to be local?  The use case I thought I was working towards was two .NET applications (one .NET 4x MVC5 the other .NET Core 2.2 Razor Pages) that had their own login UI that would utilize Owin to backend into IDSrv4 via API or function calls.

I have configured an IdentityServer4 and it appears to run properly, but this concept of UserInteraction option for Loginurl being local to IdentityServer4 vs. local to the client app was getting me stuck.  Reading this bug report/conversation, i was wondering if this feature ever got more fully fleshed out or if the intention is still that the login page is local relative to the Identity Server and that I just need to add those pages?

The use case I mention above still may be realizable, but during the Startup of my MVC5 project, it reaches out, connects with IdSrv4, and then gets to the idea that no, this user hasn't authenticated yet, so it wants to redirect to a Login screen.  I want that login screen to be for my client app, but unless I set this UserInteractionOption on the IdSrv4 app/site itself, I get default redirected to /account/login on the identity server.  

Is what I'm trying to do a reasonable use case?  I feel like this is somewhat related to another pseudo-bug report that was trying to parameterize the login redirect in some way.  I understand you have many fires to look at; if it would be possible to clarify the option var's intention and the state of the "external" vs "local" support in IdentityServer as of today (8/23/19) I'd very much appreciate it.

Thank you,

Vincent
